### PR TITLE
Ft/zenko 276 replication default read

### PR DESCRIPTION
**Pull request by alexanderchan-scality**
Review: https://github.com/scality/S3/pull/1231/commits/57a1bac41a707b22e7e71f46ae1e08c629a5df65

### CODE REVIEW

File: lib/api/apiUtils/bucket/validateReplicationConfig.js [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+ * @return {boolean} validity of replication configuration with rest endpoint
+ * configuration
+ */
+function validateConfiguration(config, bucket) {
+    const writeLocation = bucket.getLocationConstraint();
+    const readLocation = bucket.getReadLocationConstraint();
+    if (!config || !config.rules) {
+        return false;
+    }
+    const isValid = config.rules.some(rule => {
```
Comment (bennettbuchanan) [inline lib/api/apiUtils/bucket/validateReplicationConfig.js]: Don't we want to check that every rule includes the storage class?

File: lib/api/bucketPut.js [hunk 2]
```diff
@@ -18,6 +18,32 @@ config.on('location-constraints-update', () => {
...
+function _checkPreferredLocations(location, log) {
+    const retError = loc => {
+        const errMsg = 'value of the location you are attempting to set - ' +
+        `${loc} - is not listed in the locationConstraint config`;
+        log.trace(`locationConstraint is invalid - ${errMsg}`,
+          { locationConstraint: loc });
+        return { error: errors.InvalidLocationConstraint.
+          customizeDescription(errMsg) };
+    };
+    if (typeof location === 'string') {
```
Comment (bennettbuchanan) [inline lib/api/bucketPut.js]: Seems like this could be reduced to a single condition.

**Comment by alexanderchan-scality** [comment]
Added into https://github.com/scality/S3/pull/1227 as they are the same topic.

File: lib/api/apiUtils/bucket/validateReplicationConfig.js (hunk 1 above)
Comment (alexanderchan-scality) [reply lib/api/apiUtils/bucket/validateReplicationConfig.js]: My thought is that a valid configuration will contain the preferred read location as a storage class in one of the rules.

File: lib/Config.js [hunk 3]
```diff
@@ -78,12 +78,26 @@ function sproxydAssert(configSproxyd) {
...
+        assert(type === 'string' || type === 'object',
+            'bad config: each endpoint must be a string or an object');
+        if (type === 'string') {
+            assert(typeof locationConstraints[restEndpoints[r]] === 'object',
+            'bad config: rest endpoint target not in locationConstraints');
+        }
+        if (type === 'object') {
+            const { read, write } = restEndpoints[r];
+            assert(read && typeof read === 'string',
+                'bad config: invalid preferred read location');
```
[the remark this answers is not in the document]
Comment (alexanderchan-scality) [reply lib/Config.js]: will add the explicit error message

File: lib/api/apiUtils/bucket/validateReplicationConfig.js (hunk 1 above)
Comment (alexanderchan-scality) [reply lib/api/apiUtils/bucket/validateReplicationConfig.js]: I can see that there might be an error if the replication rule has a prefix.

Comment (bennettbuchanan) [reply lib/api/apiUtils/bucket/validateReplicationConfig.js]: Great! That's the kind of case I had in mind.

**Pull request closed without merging**